metadata/src/main/java/org/apache/kafka/controller/FeatureControlManager.java

```diff
@@ -304,6 +468,44 @@ public void replay(FeatureLevelRecord record) {
         }
     }
 
+    /**
+     * Allowed state changes:
+     * <li>UNINITIALIZED -> ANY</li>
+     * <li>PRE_MIGRATION -> MIGRATION</li>
+     * <li>MIGRATION -> POST_MIGRATION</li>
+     */
+    void replay(ZkMigrationStateRecord record) {
```

Comment:
I think this is way more complex than needed. All we need is something like this:
```
migrationControlState.set(record.toState());
```
plus a log4j message at INFO level.

If it's in the metadata log, then we do it. We don't second guess what's in the metadata log. The time for checking whether the state transition was legal was when we were deciding to write the record, not when we're replaying it.